google/api_core/retry.py

```diff
@@ -145,6 +145,85 @@ def exponential_sleep_generator(initial, maximum, multiplier=_DEFAULT_DELAY_MULT
         delay = min(delay * multiplier, maximum)
 
 
+def retry_target_generator(
+    target, predicate, sleep_generator, timeout=None, on_error=None, **kwargs
+):
+    """Wrap a generator object and retry if it fails.
+
+    This is the lowest-level retry helper. Generally, you'll use the
+    higher-level retry helper :class:`Retry`.
+
+    Args:
+        target(Callable[None, Generator]): A generator function to yield from.
+            This must be a nullary function - apply arguments with `functools.partial`.
+        predicate (Callable[Exception]): A callable used to determine if an
+            exception raised by the target should be considered retryable.
+            It should return True to retry or False otherwise.
+        sleep_generator (Iterable[float]): An infinite iterator that determines
+            how long to sleep between retries.
+        timeout (float): How long to keep retrying the target.
+        on_error (Callable[Exception]): A function to call while processing a
+            retryable exception. Non-None values returned by `on_error` will be
+            yielded for downstream consumers. Any error raised by this function
+            will *not* be caught.
+        deadline (float): DEPRECATED: use ``timeout`` instead. For backward
+            compatibility, if specified it will override ``timeout`` parameter.
+
+    Returns:
+        Generator: returns a generator that wraps the target in retry logic.
+
+    Raises:
+        google.api_core.RetryError: If the deadline is exceeded while retrying.
+        ValueError: If the sleep generator stops yielding values.
+        Exception: If the target raises a method that isn't retryable.
+    """
+
+    timeout = kwargs.get("deadline", timeout)
+
+    if timeout is not None:
+        deadline = datetime_helpers.utcnow() + datetime.timedelta(seconds=timeout)
+    else:
+        deadline = None
+
+    last_exc = None
+    subgenerator = None
+    for sleep in sleep_generator:
+        # Start a new retry loop
+        try:
+            # create and yeild from a new instance of the generator from input generator function
+            subgenerator = target()
+            return (yield from subgenerator)
```

Comment:
Between `retry_target` and `retry_target_generator`, what are some differences besides this?  If all the surrounding behaviors are identical, is this maybe a chance to refactor the code somewhat?

(I have the same questions about the async versions of the two functions.)

Reply:
Refactoring like that has more complications than are visible at first glance, because the presence of `yield` automatically changes a function's signature into a generator function, even if it's behind a conditional branch that's not executed. So we need different entrypoint functions for generator vs standard wrappers

I originally tried pulling out some of the shared code blocks at the beginning and end into helper functions, but a lot of context arguments were needed, and it ended up being only a couple lines that could be shared, so it felt less readable than having some duplicated structure.

Reply:
That said, looking at it again, there may be a different approach that could work, by writing a custom Generator class implementation, and returning that instead of relying on the built-in `yield`. I might give that a look to see if that can make things cleaner.